### PR TITLE
Add megaservice definition without microservice wrappers

### PR DESCRIPTION
**Pull request by Spycsh**
## Description

We will remove the microservice wrappers (e.g. for LLM, TEI) and forward the messages directly to servers. This aims to help users define a whole pipeline more easily, decrease network overhead, and also improve scaling-in scaling-out capabilities.

Users can rewrite the `align_inputs`, `align_outputs`, `align_generator` to align the format consistency between microservices.

## Issues

https://github.com/opea-project/GenAIComps/pull/582

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would break existing design and interface)
- [ ] Others (enhancement, documentation, validation, etc.)

## Dependencies

None
## Tests

Local Test

### CODE REVIEW

File: ChatQnA/docker/chatqna_no_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    for line in gen:
+        line = line.decode("utf-8")
+        start = line.find("{")
+        end = line.rfind("}") + 1
+
+        json_str = line[start:end]
+        try:
+            # sometimes yield empty chunk, do a fallback here
+            json_data = json.loads(json_str)
+            yield f"data: {repr(json_data['token']['text'].encode('utf-8'))}\n\n"
```
Comment (lkk12014402) [inline ChatQnA/docker/chatqna_no_wrapper.py:131]: should we follow openai respone format here?

Comment (Spycsh) [reply ChatQnA/docker/chatqna_no_wrapper.py:131]: Does our UI accept openai compatible response? I've tested it with our current UI and this line makes the output correctly displayed on frontend.

File: ChatQnA/docker/chatqna_no_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+import json
+import os
+import re
+
+from comps import ChatQnAGateway, MicroService, ServiceOrchestrator, ServiceType
+
+
+class ChatTemplate:
```
Comment (lkk12014402) [inline ChatQnA/docker/chatqna_no_wrapper.py:11]: how to customize template by users

File: ChatQnA/docker/chatqna_no_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        inputs["inputs"] = inputs["text"]
+        del inputs["text"]
+    return inputs
+
+
+def align_outputs(self, data, cur_node, inputs, runtime_graph, llm_parameters_dict):
+    next_data = {}
+    if self.services[cur_node].service_type == ServiceType.EMBEDDING:
+        assert isinstance(data, list)
+        next_data = {"text": inputs["inputs"], "embedding": data[0]}
```
Comment (lkk12014402) [inline ChatQnA/docker/chatqna_no_wrapper.py:67]: add retrieval params

File: ChatQnA/docker/chatqna_no_wrapper.py (hunk 2 above)
Comment (Spycsh) [reply ChatQnA/docker/chatqna_no_wrapper.py:11]: I think here we move the logics here in mega definition to let users customize the template themselves

File: ChatQnA/docker/chatqna_no_wrapper.py (hunk 3 above)
Comment (Spycsh) [reply ChatQnA/docker/chatqna_no_wrapper.py:73]: I already added the access to retrieval and rerank, inspired by https://github.com/opea-project/GenAIComps/pull/324.

**Comment by Spycsh** [comment]
Please wait for local validation.

File: ChatQnA/docker/chatqna_no_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+            endpoint="/v1/retrieval",
+            use_remote_service=True,
+            service_type=ServiceType.RETRIEVER,
+        )
+
+        llm = MicroService(
+            name="llm",
+            host=LLM_SERVER_HOST_IP,
+            port=LLM_SERVER_PORT,
+            endpoint="/generate_stream",  # FIXME non-stream case
```
Comment (lvliang-intel) [inline ChatQnA/docker/chatqna_no_wrapper.py:253]: use '/v1/chat/completions', every serving framework support this endpoint, otherwise it only works for TGI.

File: ChatQnA/docker/chatqna_no_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+            endpoint="/rerank",
+            use_remote_service=True,
+            service_type=ServiceType.RERANK,
+        )
+
+        llm = MicroService(
+            name="llm",
+            host=LLM_SERVER_HOST_IP,
+            port=LLM_SERVER_PORT,
+            endpoint="/generate_stream",  # FIXME non-stream case
```
Comment (lvliang-intel) [inline ChatQnA/docker/chatqna_no_wrapper.py:219]: use '/v1/chat/completions', every serving framework support this endpoint, otherwise it only works for TGI.

Comment (hshen14) [reply ChatQnA/docker/chatqna_no_wrapper.py:219]: if no_wrapper is good, do we need to distinguish wrapper and no_wrapper?

**Comment by Spycsh** [comment]
Note: now neuralchat on tgi gaudi does not has default template so its path to `/v1/chat/completions` fail with error `"Template error: template not found"`

File: ChatQnA/docker/chatqna_no_wrapper.py (hunk 4 above)
Comment (Spycsh) [reply ChatQnA/docker/chatqna_no_wrapper.py]: Done.

File: ChatQnA/docker/chatqna_no_wrapper.py (hunk 5 above)
Comment (Spycsh) [reply ChatQnA/docker/chatqna_no_wrapper.py]: let's keep the code backward compatible for now because there are still some paths/corner cases that I'm afraid this PR cannot cover. Also no_wrapper pattern may not apply to every microservice. Let's fully test it and then consider a replacement in the future.

**Comment by lvliang-intel** [comment]
@chensuyue,

why CI report port already allocated issue?

image

**Review by lvliang-intel — APPROVED** [review]

**Comment by chensuyue** [comment]
Please add new test scripts for the new mega service.  
And also add new images into https://github.com/opea-project/GenAIExamples/blob/main/ChatQnA/docker/docker_build_compose.yaml

**Comment by Spycsh** [comment]
> Please add new test scripts for the new mega service. And also add new images into https://github.com/opea-project/GenAIExamples/blob/main/ChatQnA/docker/docker_build_compose.yaml

Done.

**Review by chensuyue — APPROVED** [review]
